staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiserver.go

```diff
@@ -233,19 +238,28 @@ func (c completedConfig) NewWithDelegate(delegationTarget genericapiserver.Deleg
 	return s, nil
 }
 
-// AddAPIService adds an API service.  It is not thread-safe, so only call it on one thread at a time please.
-// It's a slow moving API, so it's ok to run the controller on a single thread
+// AddAPIService adds an API service. It's a slow moving API, so it's ok to acquire a lock.
 func (s *APIAggregator) AddAPIService(apiService *apiregistration.APIService) error {
 	// if the proxyHandler already exists, it needs to be updated. The aggregation bits do not
 	// since they are wired against listers because they require multiple resources to respond
-	if proxyHandler, exists := s.proxyHandlers[apiService.Name]; exists {
-		proxyHandler.updateAPIService(apiService)
-		if s.openAPIAggregationController != nil {
-			s.openAPIAggregationController.UpdateAPIService(proxyHandler, apiService)
+	updateAPIServiceFn := func() bool {
+		s.mutex.RLock()
+		defer s.mutex.RUnlock()
+		if proxyHandler, exists := s.proxyHandlers[apiService.Name]; exists {
+			proxyHandler.updateWhileKeepingAvailability(apiService)
+			if s.openAPIAggregationController != nil {
+				s.openAPIAggregationController.UpdateAPIService(proxyHandler, apiService)
+			}
+			return true
 		}
+		return false
+	}
+	if serviceUpdated := updateAPIServiceFn(); serviceUpdated {
 		return nil
 	}
 
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
 	proxyPath := "/apis/" + apiService.Spec.Group + "/" + apiService.Spec.Version
```

Comment:
should we try updateAPIServiceFn again here with the write lock? Can there be a race otherwise?

Reply:
No, `AddAPIService` is thread-safe. The function is called from controller (single-threaded)  https://github.com/kubernetes/kubernetes/blob/6e31183d4780ec340af59c333dca6163ac79ad18/staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiservice_controller.go#L86 
So if the service does not exist we acquire the lock and add it to the map. I'll document this behavior.